### ISSUE TITLE
Semi-random test failure of QTable operations

### ISSUE THREAD

**Issue by bsipocz** [astropy/astropy#6291]
This failure shows up relatively randomly, usually either in the python3.4+np1.9 or python3.5+np.1.11 travis job, but I saw it locally, too with python3.5+np1.12
 https://travis-ci.org/astropy/astropy/jobs/247412542#L2795

@mhvk @taldcroft - Is it real, can we do something about it?

**Comment by pllim** [comment]
`git blame` in `astropy/units/quantity.py` around the line of interest points to maybe #6137? The other commits are much older.

**Comment by mhvk** [comment]
Seems somewhat unlikely. Given the sporadic nature, I worry more about the change of `info._parent` to a weak reference - maybe sometimes the reference has disappeared. @taldcroft?

**Comment by bsipocz** [comment]
now it seems to be disappeared from master PR runs, but was there a few hours ago with no related changes merged since...

**Comment by taldcroft** [comment]
I have played with this a little.  I ran the tests 20+ times locally (macosx + py3.5 + numpy 1.11 / 1.12) and could not reproduce the problem.   It is a bit perplexing because (to first order) if there is a problem related to #6277 we should be seeing the `AttributeError` that comes if you access `col.info` somewhere and that is a dead weak reference.  The try/except in the code is not catching `AttributeError`.

**Comment by mhvk** [comment]
Very weird.

**Comment by bsipocz** [comment]
I've just restarted a travis job on v2.0.x branch on py3.4+np1.9 with the failure.

**Comment by olebole** [comment]
I can __confirm__ this failure with my first test build on Debian sid with Python 3.5.3. Test environment as reported in the log:

 * Numpy: 1.12.1
 * Scipy: 0.18.1
 * Matplotlib: 2.0.0
 * h5py: 2.7.0
 * Pandas: not available
 * Cython: 0.25.2
 * Using Astropy options: remote_data: none.

**Comment by astrofrog** [comment]
Just a wild guess - with weak references, if the weak reference is the last reference to an object, the exact point where the object will be garbage-collected is non-deterministic, so this might be what is happening here? (i.e. something is relying on the attribute existing but it is sometimes getting garbage collected too soon?)

**Comment by astrofrog** [comment]
One way to test my hypothesis - when creating a weakref, you can add a callback function when the original object is deleted, so this would help find out when exactly this happens and how this correlates with the test failures.

**Comment by taldcroft** [comment]
Hmm, I ran 20 times on linux (CentOS-6) and could not generate the error:
```
platform linux -- Python 3.5.1, pytest-3.1.2, py-1.4.34, pluggy-0.4.0

Running tests with Astropy version 3.0.dev19488.
Running tests in lib.linux-x86_64-3.5/astropy/table docs/table.

Date: 2017-06-28T07:34:47

Platform: Linux-2.6.32-696.1.1.el6.x86_64-x86_64-with-centos-6.9-Final

Executable: /export/tom/anaconda/envs/astropy3-1.2rc1/bin/python

Full Python Version: 
3.5.1 |Continuum Analytics, Inc.| (default, Dec  7 2015, 11:16:01) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Numpy: 1.12.1
Scipy: 0.19.0
Matplotlib: 2.0.2
h5py: not available
Pandas: not available
Cython: 0.24
Using Astropy options: remote_data: none.
```

**Comment by bsipocz** [comment]
Hmm. Is there anything we could add to the current tests to help the debugging using travis? It's appears there quite often. The latest is here: https://travis-ci.org/astropy/astropy/jobs/247743631

**Comment by mhvk** [comment]
If @olebole gets it on debian, I should be able to reproduce it too; I'll try once I get in to work.

**Comment by taldcroft** [comment]
Worst case is that we revert #6277 and then put out a bug-fix patch as soon as we can (potentially using the strategy from #6279).  The memory leak has been there since 1.1.

**Comment by mhvk** [comment]
> If @olebole gets it on debian, I should be able to reproduce it too; I'll try once I get in to work.

Well, "should" was the operative word. I cannot reproduce it...

**Comment by astrofrog** [comment]
I think there is a way to run a Travis environment (Ubuntu-based) with docker?

**Comment by taldcroft** [comment]
I remember that once before there was a weird failure that only showed up when all of the tests were being run instead of a subset. I was only running the table tests.  @mhvk?

**Comment by bsipocz** [comment]
I just had one locally when only running the table tests. Second time around it was gone again.

**Comment by taldcroft** [comment]
Can you back out 6277 locally and see if that seems to fix the problem?

**Comment by bsipocz** [comment]
@taldcroft - I've pushed a test branch with it and one without and test it on travis, as locally isn't not failing again. I'll aim to restart those test builds during tonight as many times as I can, and will report back.

**Comment by bsipocz** [comment]
this is with: https://travis-ci.org/bsipocz/astropy/builds/248049549

this is without: https://travis-ci.org/bsipocz/astropy/builds/248049612

**Comment by saimn** [comment]
I tried to understand as I came across this while running tests with pytest directly. So what I noticed is that it always fails with `pytest astropy/table/tests/test_operations.py --pdb -sv -k test_col_meta_merge_inner` but not with `python setup.py test` hum hum...
 
I don't understand all the magic with `.info`, but had another idea: it may be related to another magic with pytest's assert statement. And with the simple following change, it does not fail anymore (locally, [build](https://travis-ci.org/saimn/astropy/builds/250129876) on my going is ongoing):
```diff
--- a/astropy/table/tests/test_operations.py
+++ b/astropy/table/tests/test_operations.py
@@ -682,7 +682,8 @@ class TestVStack():
         else:
             assert len(warning_lines) == 0
             # Check QTable correctly dealt with units.
-            assert out.pformat() == ['   a       b   ',
+            outf = out.pformat()
+            assert outf == ['   a       b   ',
                                      '   km          ',
                                      '-------- ------',
                                      '0.000000    foo',
```

**Comment by saimn** [comment]
Ok it still fails, sorry for the noise...

**Comment by taldcroft** [comment]
If anyone can provide guidance or a way for me to reproduce this on a debian-like system, I could take a crack at debugging.  I found some instructions for installing the travis-CI env. with docker, but then discovered that the latest mac docker requires at least 10.10 and I'm on 10.9...  I suppose I could upgrade, but I am worried about issues with APFS or other random things breaking.

**Comment by bsipocz** [comment]
Sadly I even failed with trying to reproduce it on travis. It still shows up randomly though.

**Comment by astrofrog** [comment]
@taldcroft - I also can't reproduce the issue but I was looking at the code and one thing confused me. You seem to refer a lot to ``col.info.parent_table`` but the following properties:

```

    @property
    def parent_table(self):
        if self._parent_table is None:
            return None
        else:
            return self._parent_table()

    @parent_table.setter
    def parent_table(self, table):
        if table is None:
            self._parent_table = None
        else:
            self._parent_table = weakref.ref(table)
```

are defined on ``BaseColumn``, **not** ``BaseColumnInfo``, so I'm not sure if it ever has an effect. Did you mean to define these on ``BaseColumnInfo``? Could this be related to the failures we are seeing?

**Comment by astrofrog** [comment]
@saimn - since you are able to reproduce the issue, could you try and run the tests in the ``debugging`` branch in my fork?

**Comment by taldcroft** [comment]
@astrofrog - this part of the code is perhaps more twisted than ideal, but [here](https://github.com/astropy/astropy/blob/3973f03e366fa2daaae5750195f8dc90694800e2/astropy/table/column.py#L152) the `ColumnInfo` class specifies that `ColumnInfo` attributes are just refs to the parent `Column` attributes.  So `col.info.parent_table` resolves as `col.parent_table`.

**Comment by astrofrog** [comment]
@taldcroft - ah I see, but that then doesn't resolve for mixin columns, correct? (I think that's fine, just checking I understand)

**Comment by saimn** [comment]
@astrofrog - I was able to reproduce, with the pytest command mentioned above, but now it doesn't fail !

**Comment by astrofrog** [comment]
@saimn - you mean it doesn't fail with master or with my branch?

**Comment by saimn** [comment]
@astrofrog - with master (which I updated this morning).

**Comment by astrofrog** [comment]
![64900395](https://user-images.githubusercontent.com/314716/27918490-0b89efaa-6267-11e7-8810-e9d45d85bf59.jpg)

**Comment by bsipocz** [comment]
@astrofrog - If your fix doesn't break anything, what about getting it in? Worst case nothing will change, but maybe we never see these issues on travis ever again.

**Comment by astrofrog** [comment]
@bsipocz - my 'fix' is not really a fix and should definitely not be merged.

**Comment by taldcroft** [comment]
Plan is to back this out for 2.0 and come back to this soon after release.  I will issue a PR shortly.

**Comment by astrofrog** [comment]
I'm re-opening this for two reasons:

* We just saw the failure again when building/testing the wheels
* I just searched through all the Travis logs and so far the oldest mention of this failure I found was 2017-05-11T10:10:35Z, which is **before** https://github.com/astropy/astropy/pull/6277 was merged - I'm going to continue searching through older logs and see if I can pinpoint more precisely when it started happening.

**Comment by astrofrog** [comment]
It looks like this matches up nicely with https://github.com/astropy/astropy/pull/5841, which adds the actual test code that is failing sporadically! So it seems the bug was in there from the start and NOT introduced by the weakref changes (which I think can be added back as-is).

Since the weakref is not at play, the only other source of randomness I can think of is related to e.g. sets or dictionaries being unordered and occasionally ending up in an order that causes issues.

**Comment by astrofrog** [comment]
I have been debugging this a bit by adding some print statements to the test and then running on travis until it fails. What I've found so far is that I already get a failure if I try and do:

```
        print('table 1')
        print(type(t1))
        print(t1)
```

**before** the vstack command, and this outputs:

```
table 1
<class 'astropy.table.table.QTable'>
```

then crashes in the usual way. I'll be adding more here as I inspect t1 in more detail.

**Comment by astrofrog** [comment]
Here's some more debugging info. Incidentally this is the most inefficient debugging, ever!

```
table 1
type(t1) <class 'astropy.table.table.QTable'>
type(t1['a']) <class 'astropy.units.quantity.Quantity'>
type(t1['b']) <class 'astropy.table.column.Column'>
type(t1['a'].info) <class 'astropy.units.quantity.QuantityInfo'>
type(t1['b'].info) <class 'astropy.table.column.ColumnInfo'>
t1['a'].info:
name = a
dtype = float64
unit = cm
format = %f
class = Quantity
n_bad = 0
length = 2
t1['b'].info name = b
dtype = str3
description = t1_b
class = Column
n_bad = 0
length = 2
t1['a'].info.format %f
t1['b'].info.format None
t1['a'].info.possible_string_format_functions <function QuantityInfoBase.possible_string_format_functions at 0x7f488e7709d8>
t1['b'].info.possible_string_format_functions: no attribute
```

**Comment by mhvk** [comment]
Looking at this because of #6725 - where the `xfail` on the tests might get reverted. Could this be the same bug that bit @larrybradley (#6703), where there was sometimes a reuse of the format function. If so, then this should be solved by #6714.

**Comment by bsipocz** [comment]
We don't see this issue for a while, so the fix mention above seem to be working. I'm closing it now, but reopen if it pops up again.
